### ISSUE TITLE
Allow relative dates for default values in SQL variables

### ISSUE THREAD

**Issue by mazameli** [metabase/metabase#4509]
Per a [Discourse discussion](http://discourse.metabase.com/t/defaulting-dates-in-sql-questions/1250). Right now you can only select a single date, though there are lots of times where you'd want the default to be "today" or "last 30 days":
![screen shot 2017-03-09 at 12 24 49 pm](https://cloud.githubusercontent.com/assets/2223916/23769210/86c0f9ae-04c3-11e7-8e89-b998c5ddb1ac.png)

I suppose the challenge is that many of our relative date selections are actually `between` clauses, meaning you're actually mapping to two variables. There are some exceptions, such as "today," but not sure how we'd do "Last 30 days" with our current SQL template tags.

**Comment by tanzoniteblack** [comment]
The date SQL variable only allows you to select a single date anyways, so if we show only relative dates that are single values rather than ranges (i.e. `today` or `yesterday` but not `last week`, `last 30 days`, etc.) then it would probably be "good enough".

**Comment by moumny** [comment]
Creat design ! I like the "required" field very much.
Another option that could be convienient is to set "maximum date range" eg. maximum one month to avoid having users querying too much data.

**Comment by tlrobinson** [comment]
I believe this is resolved. @mazameli can you verify?

**Comment by mazameli** [comment]
Sort of. I can only get it to work with field filter variables (expected?).  And when trying to run:

SELECT count(*)
FROM products
where CREATED_AT = {{t}}

I get `Cannot parse "TIMESTAMP" constant "1"` with these variable settings:

![screen shot 2017-05-03 at 11 19 25 am](https://cloud.githubusercontent.com/assets/2223916/25675175/62b601ee-2ff2-11e7-8627-d2e3b4790240.png)

**Comment by mazameli** [comment]
Sort of. I can only get it to work with field filter variables (expected?).  And when trying to run:

SELECT count(*)
FROM products
where CREATED_AT = {{t}}

I get `Cannot parse "TIMESTAMP" constant "1"` with these variable settings:

![screen shot 2017-05-03 at 11 19 25 am](https://cloud.githubusercontent.com/assets/2223916/25675175/62b601ee-2ff2-11e7-8627-d2e3b4790240.png)

**Comment by mazameli** [comment]
I think I'm doing this wrong, actually. It should be `where {{t}}`, which works.

**Comment by mazameli** [comment]
But, again, I can't recall if the expected behavior is to have the same relative options for default values for variables of type `Date`, not just for `Field Filters`

**Comment by tlrobinson** [comment]
> I can only get it to work with field filter variables (expected?)

Yes, non field filter date variables can't do ranges, only a single value.

We could (in theory) support relative dates, and "30 days ago" would substitute `now() - interval '30 days'` or whatever. Not sure how useful that is.

**Comment by mazameli** [comment]
> Yes, non field filter date variables can't do ranges, only a single value.

Yeah, I just wasn't sure if we were trying to support `yesterday`, `today`, `tomorrow` for example. I think that was @tanzoniteblack's main usecase:

> The date SQL variable only allows you to select a single date anyways, so if we show only relative dates that are single values rather than ranges (i.e. today or yesterday but not last week, last 30 days, etc.) then it would probably be "good enough".

**Comment by mazameli** [comment]
Though I guess it doesn't matter really since you can do `yesterday` etc. with field filters.

**Comment by tanzoniteblack** [comment]
The new filters that allow us to use relative dates are nice, and solve a section of the use case I'd intended.

However, many of the times I want to use these filters, I really only want to set the default to `Today` and still have the normal calendar widget for the user to interact with. I'm not sure why it's required that we when creating the question define which widget for time selection is required when doing a field method, rather than having a "Time Widget" that allows the user when using the question to select for themselves whether they want a single specific date, a span, or a relative time period.

**Comment by salsakran** [comment]
I think the right solution here would be to allow Field-base-filter-clauses to display widgets on the report itself.

**Comment by salsakran** [comment]
bumping this to 0.25 to give us a bit of room to figure out how to approach this. specifically, we should unify how question params work with user-space chrome (eg, the granularity controls + filters in the new QB timeseries mode), GUI query parameters (#4716), and dashboard filters. Right now all three are subtly different.

**Comment by bendem** [comment]
Hi there, I just found this issue and I think my problem fits the current discussion.
Many of our queries are built build joining our tables to 
```sql
from generate_series(
    date_trunc('day', {{from}}),
    date_trunc('day', {{to}}),
    interval '1 day'
) series(day)
```

From a user PoV, I'd love to be able to have something like
```sql
from generate_series(
    date_trunc('day', {{range.start}}),
    date_trunc('day', {{range.end}}),
    interval '1 day'
) series(day)
```
and mark the range variable as a date range.

**Comment by ConradIrwin** [comment]
I'd like to second the ability to have relative dates in normal filters.

Our use-case is for parameterizing bigquery's _PARTITIONTIME field, which metabase does not expose as a field.

**Comment by Giannettig** [comment]
Hi guys, I need to filter today AND tomorrow in the query and it seems to me that I can either filter tomorrow OR today but not both. I tried selecting next one day  and checkboxing current day and it di not work.

**Comment by jbrooksuk** [comment]
I have an example where a Field Filter doesn't cut it due to a slight complexity in the query, but where we'd want to use a relative date.

```sql
AND (
        (DATE(date_completed) BETWEEN {{start_date}} AND {{end_date}})
        OR (DATE(date_booked) BETWEEN {{start_date}} AND {{end_date}})
        )
```

Because of the **OR**, we can't use a Field Filter (unless I'm missing something).

**Comment by flamber** [comment]
I think this might be a duplicate of #3156 - or at least very related.

**Comment by HKFiliatix** [comment]
Any updates? What is the problem to add this even on the front side? Make just widget button which will  return new Date() as value? I try to use question in pulse, but filter haven't 'today' value!

**Comment by uhrohraggy** [comment]
This is still an issue/feature request. I would like to at least see an option for the default value be today. setting a value of Oct 15, 2021 doesn't help when this query is run 2 years from now and still shows Oct 15, 2021.

**Comment by SCT-Shimazaki** [comment]
I want this feature too, preferably with options like "default to yesterday".
 In my case I am writing a mongo query and since I cannot do calculation like - (minus) or multiplication (*), I am forced to use a date picker like this.... 
I thought of using Redash for the purpose but it turned out Redash requires a bit too much resources for my use case, and the UI was more confusing.

**Comment by xujijun** [comment]
Hello, any plan on this issue :)

**Comment by damaon** [comment]
I used such workaround for it:
```
CREATE VIEW `VIEW_timestamps_daily` AS
SELECT TIMESTAMP(day)
FROM UNNEST(GENERATE_DATE_ARRAY('2024-01-01', CURRENT_DATE())) AS day;
```
And then:
```
WITH DateRange AS (
SELECT MIN(timestamp) as start_date, MAX(timestamp) as end_date
FROM `VIEW_timestamps_daily`
WHERE {{date_filter}}
)
```
and
```
SELECT ... WHERE date > (SELECT start_date FROM DateRange) ... 
```

**Comment by johnsmclay** [comment]
Not having a relative option on single date filters makes them almost useless for dashboards.  The cases for a dashboard or any kind of standard report that is ALWAYS a specific date from the past (i.e. every day checking how I compare to the 15th of April 2024, or what the results from that date were) are practically nil.  Even for historical fact-finding, "one week ago", "3 days ago", etc are going to be way better than just pick a date from a calendar.  ESPECIALLY for defaults.
